source/common/common/perf_tracing.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#ifdef ENVOY_PERFETTO
+
+#include "perfetto.h"
+
+// `Perfetto` is an open-source stack for performance instrumentation and trace
+// analysis. In Envoy we use it only as a library for recording app-level
+// traces which can be later analyzed online at https://ui.perfetto.dev/ or wit
+// custom tools.
+//
+// The support is enabled with
+//   bazel --define=perf_tracing=enabled ...
+// In the absence of such directives the macros for instrumenting code for
+// performance analysis will expand to nothing.
+//
+// The supported `Perfetto` macros are TRACE_EVENT, TRACE_EVENT_BEGIN,
+// TRACE_EVENT_END and TRACE_COUNTER.
+//
+// See https://perfetto.dev/docs/instrumentation/track-events for more details.
+
+// NOLINT(namespace-envoy)
+
+PERFETTO_DEFINE_CATEGORIES(
+    perfetto::Category("core").SetDescription("Events from core modules"),
+    perfetto::Category("extensions").SetDescription("Events from extensions"));
+
+#else
+
+// Macros that expand to nothing when performance collection is disabled. These are contrived to
+// work syntactically as a C++ statement (e.g. if (foo) TRACE_COUNTER(...) else TRACE_COUNTER(...)).
+
+#define TRACE_EVENT(category, name, ...)                                                           \
+  do {                                                                                             \
+  } while (false)
+#define TRACE_EVENT_BEGIN(category, name, ...)                                                     \
+  do {                                                                                             \
+  } while (false)
+#define TRACE_EVENT_END(category, ...)                                                             \
+  do {                                                                                             \
+  } while (false)
+#define TRACE_COUNTER(category, track, ...)                                                        \
+  do {                                                                                             \
+  } while (false)
```

Comment:
Per earlier comment I think it'd be good to understand when to use the existing source/common/common/perf_annotation.h and when to use this, or, assuming this is a lot better, remove perf_annotation.h and replace calls to it existing in the codebase with calls to this.

Reply:
I added a couple of macros to `ActiveStream` - one scoped macro and a pair of macros to trace an object's life span - as an illustration.

I think the purposes of perf annotations and this perf tracing are different a bit. The former works great for benchmarking, the latter gives better insight on what happens in what order and how much time it takes. Also its cross-scoped macros don't work well for open-loop benchmarks (`TRACE_EVENT_BEGIN` for a single category can't be nested in a single track). I'd keep both.

Reply:
OK, that's fine. Can we, in this PR, put comments in both headers to xref one another and describe when you'd use one vs the other? Just from the naming of them it's hard to discern when to use each one.

Reply:
I guess I'm not fully convinced yet that perf annotations can't be replaced with tracing. @rojkov can you explain the open loop benchmark situation further?

Reply:
In open-loop benchmarking new requests are fired regardless of completion of any requests sent previously. For example two requests are sent one after another, but the second one is sent before the first response is received.

Now consider putting `TRACE_EVENT_BEGIN("core", "ActiveStreamLifeSpan")` into `ActiveStream`'s constructor and `TRACE_EVENT_END("core")` into its destructor. The first request will emit one opening event for "ActiveStreamLifeSpan". The second request will emit one more opening event. Now there are two opening events in the stack. The response for the first request will emit the first closing event. Here's the problem: it will close the most recent opening event in the stack which corresponds to the second request. And BTW it will close the most recent opening event irrespective of its name. To avoid that the corresponding opening and closing events should be placed on separate tracks in addition to the existing thread tracks. In this particular case every new request creates at least one new track. Benchmarks may create hundreds or thousands of such tracks. I use https://ui.perfetto.dev/ for analysis and it simply doesn't show anything for trace files with so many tracks. And if it did such a visualization would be too cluttered and useless.

Indeed this issue could be circumvented with handcrafted SQL queries, I guess, or even with custom trace-based metrics added to the Trace Processor, but perf annotations "just work" in this case.

I'm going to learn more about extending the Trace Processor with custom metrics, but until then dropping perf annotations would be premature I guess.

Reply:
Thanks for the detailed explanation. I agree, this seems broken and the basic issue here is that (correct me if I'm wrong) we're an event-based system and the trace framework isn't tracking context across events on the same thread. If we try and teach it this, it produces a large number of independent traces that the tooling doesn't do a good job at synthesizing. Is this the gist?

If so, I'm wondering how this is ever useful in an Envoy worker thread. The problem doesn't seem confined to open-loop benchmarking. What's a situation in which the cross-request tracking doesn't act in a misleading way?

Reply:
You understood it correctly.

Perfetto produces easily consumable data for cross-stack events in case of consecutive requests produced by one virtual user (`--concurrency=1` in terms of nighthawk). If we teach Perfetto to put such events on separate tracks the produced data would not be misleading still, but it would require additional manipulations to extract useful bits. This is my current understanding.

Perf annotations do a similar job for cross-stack perf operations: every such perf operation is associated with a class instance which owns it. Perfetto events can be associated with `this` too via a track specially minted for `this`. The difference is perf annotations ignore these owner-operation relationships in the final report automatically. Perfetto keeps this info requiring additional post-processing of the tracing data.

If we forget about cross-stack operations/events and use RAII-based ones only then Perfetto provides better visualization out of the box.

Also for envoy-mobile purposes the tracing info generated by Perfetto seems to be more useful.

Reply:
OK, this is great. Can you somehow capture this in a form that we can make referencable the next time this discussion comes up? Maybe a dedicated issue or add to the usage docs (I know there is already track mention there, but it's hard to infer the relationship between different tools from that and when to use one vs. the other).